bit9PlatformAPI/examples/connectors/Metascan.py

```diff
@@ -0,0 +1,247 @@
+        n = 0
+        for key in scans:
+            s = scans[key]
+            if s['scan_result_i']:
+                if n != 0:
+                    notification['malwareType'] += '; '
+                notification['malwareName'] += '; '
+                notification['malwareType'] += key + ':' + s['threat_found']
+                notification['malwareName'] += s['threat_found']
+                n += 1
+# Let's see if it is malicious. Use some fancy heuristics...
+        positivesPerc = 100 * n / self.msavtotal
+        if positivesPerc > 50:
+            notification['analysisResult'] = 3  # ...malicious
+            notification['severity'] = 'critical';
+            notification['type'] = 'malicious_file';
+        elif positivesPerc > 0:
+            notification['analysisResult'] = 2  # ...could be malicious
+            notification['severity'] = 'high';
+            notification['type'] = 'potential_risk_file';
+        else:
+            notification['analysisResult'] = 1  # clean!
+            notification['severity'] = 'low';
+            notification['type'] = 'clean_file';
+        # Send notification
+        self.bit9.create("v1/notification", notification)
+        print("MS analysis for fileAnalysis %d completed. MS result is %d%% malware (%s). Reporting status: %s" % (fileAnalysisId, positivesPerc, notification['malwareName'], notification['type']))
+
+    def processOneAnalysisRequest(self, pa):
+        # Use md5 hash if we have one. If not, use Sha256
+        fileHash = pa['md5'].strip()
+        if fileHash == '':
+            fileHash = pa['sha256'].strip()
+
+        lastAttempt = None
+        dataId = None
+        # Check our cache if we already sent this file for scan
+        if fileHash in self.scheduledScans.keys():
+            lastAttempt = self.scheduledScans[fileHash]
+            # Be polite and don't keep asking MS for status too often. If we already tried recently, bail
+            if lastAttempt['nextCheck'] > datetime.datetime.now():
+                return
+            # Get our dataId we got from MS last time around
+            print("Now attempting to send the data ID")
+            dataId = lastAttempt['data_id']
+            r = requests.get(self.ms_url + "/file/" + dataId, params={'':''})
+            print("This is the URL being sent %s" % (self.ms_url + "/file/" + dataId))
+            print(r)
+        else:
+            # we have not asked MS yet. Try with file hash
+            r = requests.get(self.ms_url + "/hash/" + fileHash, params={'':''})
+
+        r.raise_for_status()
+        if r.status_code == 204:
+            # no results from MS because rate limit was reached. We will try again later.
+            return
+
+        scanResults = r.json()
+        x=scanResults.get("scan_results",{})
+        # Check if we got results...
+        self.msresults=x.get('scan_all_result_i')
+        self.msavtotal=x.get('total_avs')
+        if self.msresults is not None:
```

Comment:
I think I'm going to need help with that. I went though this with Dan Moss from Bit9 and this is what we came up with. I can ask our resident Python guy to help me with it. Maybe he can help me scope it down.
-Patrick